natlas-server/entrypoint.sh

```diff
@@ -1,6 +1,14 @@
 #!/bin/sh
 
-mkdir -p /data/{db,media}
+mkdir -p /data/{db,media} logs
+
+DATABASE_FILE=${SQLALCHEMY_DATABASE_URI//sqlite:\/\/\//}
```

Comment:
What is this doing? `sqlite:\/\/\//` is missing an escaped slash, but even then I'm not really sure what the intention is. If you don't mount the data directory, then other things will not be persistent anyways, and if you do mount the data directory, then the Dockerfile defines the `SQLALCHEMY_DATABASE_URI` inside the persistent data directory. From what I can divine this is doing, you'd be creating the database in an undefined location, if it works at all (I can't find any documentation suggesting that is a valid sqlite URI), which I would guess is the current working directory, which would be in non-persistent space? I've never seen this syntax before though, so I am not really sure what it's doing.

Reply:
>The path component of the URI specifies the disk file that is the SQLite database to be opened. If the path component is omitted, then the database is stored in a temporary file that will be automatically deleted when the database connection closes.

https://www.sqlite.org/uri.html

`sqlite:////` means that when you run `flask db upgrade` you are creating a temporary file, and then when flask db upgrade is finished and the app shuts down, that file is deleted.

Reply:
Search https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameter-Expansion for ${parameter/pattern/string}

It does a search and replace via bash magic. it takes the sqlite:/// in sqlite:////data/db/metadata.db and replaces it with "" making it so that DATABASE_FILE is /data/db/metadata.db. Then it checks to see if it exists. If it doesn't it will run the ` flask db upgrade`. 

I do mount this data dir in via the docker-compose.yaml. ns-data is a docker persistent volume. 
```
    volumes:
      - ns-data:/data/
      - ${PWD}/myscopefile.txt:/myscopefile.txt
```

Without this it would not have a db/tables and it would throw all kinds of errors on start.

Reply:
Dangit I typed out a whole long thing and then hit resolve and it erased all my content.

Okay I get what the search and replace is doing. It's ugly, but that's to be expected of bash I suppose. I might just add a comment to the code to make it more clear, since I don't know how common that syntax is.

Another trouble is that merely checking the existence of the metadata.db file isn't the only time that `flask db upgrade` needs to run. It also needs to run whenever a new docker container is downloaded and there is a new schema migration. Technically speaking, there is no harm in running `flask db upgrade` on every container start. When there are no new migrations, `flask db upgrade` is effectively no-op. It checks the migrations folder for the latest migration, and then it checks the schema, and if they match then it exits. If they don't match, then it runs migrations until the schema is brought in line with the HEAD of migrations. Running `flask db upgrade` every time the container starts is kind of... idk, it seems wrong I guess, but I can't think of a better way to do it that scales out to also supporting non-sqlite deployments (which may become more common as the database becomes used for more things).